worstcase/state.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// Package worstcase provides the worst-case balance and nonce tracking needed
+// to safely include transactions that are guaranteed to be valid during
+// execution.
+package worstcase
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core"
+	"github.com/ava-labs/libevm/core/state"
+	"github.com/ava-labs/libevm/core/txpool"
+	"github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/params"
+	"github.com/holiman/uint256"
+
+	"github.com/ava-labs/strevm/blocks"
+	"github.com/ava-labs/strevm/gastime"
+	"github.com/ava-labs/strevm/hook"
+	saeparams "github.com/ava-labs/strevm/params"
+)
+
+// State tracks the worst-case gas price and account state as operations are
+// executed.
+//
+// Usage of the [State] should follow the pattern:
+//  1. [State.StartBlock] for each block to be included.
+//  2. [State.GasLimit] and [State.BaseFee] to query the block's parameters.
+//  3. [State.ApplyTx] or [State.Apply] for each transaction or operation to
+//     include in the block.
+//  4. [State.FinishBlock] to finalize the block's gas time.
+//  5. Repeat from step 1 for the next block.
+type State struct {
+	hooks  hook.Points
+	config *params.ChainConfig
+
+	db    *state.StateDB
+	clock *gastime.Time
+	// expectedParentHash is used to sanity check that blocks are provided
+	// in-order.
+	expectedParentHash common.Hash
+
+	qSize, blockSize, maxBlockSize gas.Gas
+
+	baseFee *uint256.Int
+	curr    *types.Header
+	rules   params.Rules
+	signer  types.Signer
+}
+
+var errSettledBlockNotExecuted = errors.New("settled block not executed")
+
+// NewState constructs a new worst-case state on top of the settled block.
+func NewState(
+	hooks hook.Points,
+	config *params.ChainConfig,
+	stateCache state.Database,
+	settled *blocks.Block,
+) (*State, error) {
+	if !settled.Executed() {
+		return nil, errSettledBlockNotExecuted
+	}
+
+	// TODO: Should we be providing snapshots here?
+	db, err := state.New(settled.PostExecutionStateRoot(), stateCache, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &State{
+		hooks:              hooks,
+		config:             config,
+		db:                 db,
+		clock:              settled.ExecutedByGasTime(),
+		expectedParentHash: settled.Hash(),
+	}, nil
+}
+
+const rateToMaxBlockSize = saeparams.Tau * saeparams.Lambda
+
+var (
+	errNonConsecutiveBlocks = errors.New("non-consecutive blocks")
+	errQueueFull            = errors.New("queue full")
+)
+
+// StartBlock updates the worst-case state to the beginning of the provided
+// block.
+//
+// It is not necessary for [types.Header.GasLimit] nor [types.Header.BaseFee] to
+// be set. However, all other fields should be populated and
+// [types.Header.ParentHash] must match the previous block's hash.
```

Comment:
> all other fields should be populated

I say this because we end up passing this header into `hooks.SubSecondBlockTime`.